### PR TITLE
[RSDK-5431] Properly handle NaN values being passed into motion MoveOnGlobe

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/RSDK-5431)

- Update the slam & movement sensor localizers to return errors if any of their outputs would contain `NaN`
- Change `motion.NewMovementSensorLocalizer` to return an error if any of its parameters contained `NaN`
- Change `defaultReplanCostFactor` from `1.0` to `2.0`
- Change `utils.NewFilePathDebugLogger` to log at debug level
- Add `utils.SwapCompasHeadingHandedness`
- Add tests for all new behavior

### CODE REVIEW

File: services/motion/builtin/move_request.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ import (
 )
 
 const (
-	defaultReplanCostFactor = 1.0
+	defaultReplanCostFactor = 2.0
```
Comment (nicksanford) [inline services/motion/builtin/move_request.go:28]: This was necessary to get MoveOnGlobe to work at all during outdoor hardware testing. Do we want to make this change here or in a different PR? Does this require a ticket?

File: services/motion/builtin/move_request.go [hunk 2]
```diff
@@ -314,12 +313,23 @@ func (ms *builtIn) newMoveOnGlobeRequest(
...
+	ms.logger.Debugf("move on globe origin: %v", origin)
+	if err := motion.ValidateGeopoint(origin); err != nil {
+		return nil, err
+	}
+
+	heading, err := movementSensor.CompassHeading(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	valExtra.extra["origin"] = *spatialmath.NewGeoPose(origin, heading)
```
Comment (nicksanford) [inline services/motion/builtin/move_request.go:325]: remove

File: services/motion/localizer.go [hunk 3]
```diff
@@ -66,20 +125,29 @@ func (m *movementSensorLocalizer) CurrentPosition(ctx context.Context) (*referen
 	if err != nil {
 		return nil, err
 	}
+
+	if err := ValidateGeopoint(gp); err != nil {
```
Comment (nicksanford) [inline services/motion/localizer.go:129]: test for this

File: services/motion/localizer.go [hunk 4]
```diff
@@ -56,8 +61,62 @@ type movementSensorLocalizer struct {
...
+	if math.IsNaN(gp.Lng()) {
+		return errors.New("lng can't be NaN")
+	}
+
+	return nil
+}
+
+// ValidatePose validates that a pose can be used for
+// motion planning.
+func ValidatePose(p spatialmath.Pose) error {
```
Comment (nicksanford) [inline services/motion/localizer.go:92]: turns out that if any of the X, Y, Z of the R3.Vector passed to create the pose are NaN X,Y,Z of the resulting pose will all be nil so in practice I've only been able to get this first error to trigger.

File: services/motion/localizer.go [hunk 5]
```diff
@@ -90,5 +158,10 @@ func (m *movementSensorLocalizer) CurrentPosition(ctx context.Context) (*referen
 	}
 
 	pose := spatialmath.NewPose(spatialmath.GeoPointToPose(gp, m.origin).Point(), o)
-	return referenceframe.NewPoseInFrame(m.Name().Name, spatialmath.Compose(pose, m.calibration)), nil
+	pif := referenceframe.NewPoseInFrame(m.Name().Name, spatialmath.Compose(pose, m.calibration))
+	if err := ValidatePose(pif.Pose()); err != nil {
```
Comment (nicksanford) [inline services/motion/localizer.go:162]: figure out how to trigger this

File: utils/math.go [hunk 6]
```diff
@@ -16,6 +16,12 @@ func DegToRad(degrees float64) float64 {
 	return degrees * math.Pi / 180
 }
 
+// SwapCompasHeadingHandedness converts left hand rule compass heading degrees
+// to right hand rule compass heading degrees & vice versa.
+func SwapCompasHeadingHandedness(heading float64) float64 {
```
Comment (biotinker) [inline utils/math.go:21]: Compass has two 's's

Comment (nicksanford) [reply utils/math.go]: thanks! Fixed

File: components/base/kinematicbase/ptgKinematics.go [hunk 7]
```diff
@@ -223,6 +223,10 @@ func (ptgk *ptgBaseKinematics) ErrorState(ctx context.Context, plan [][]referenc
 		return nil, err
 	}
 	nominalPose = spatialmath.Compose(runningPose, currPose)
+	ptgk.logger.Debugf("curr position %#v", spatialmath.PoseToProtobuf(actualPIF.Pose()))
```
Comment (biotinker) [inline components/base/kinematicbase/ptgKinematics.go:226]: I'm not sure we want these to be actually merged into main

Comment (nicksanford) [reply components/base/kinematicbase/ptgKinematics.go]: removed

File: services/motion/localizer.go [hunk 8]
```diff
@@ -66,20 +125,29 @@ func (m *movementSensorLocalizer) CurrentPosition(ctx context.Context) (*referen
...
 			return nil, err
 		}
+
+		if math.IsNaN(headingLeft) {
+			return nil, errors.New("heading can't be NaN")
+		}
 		// CompassHeading is a left-handed value. Convert to be right-handed. Use math.Mod to ensure that 0 reports 0 rather than 360.
-		heading := math.Mod(math.Abs(headingLeft-360), 360)
-		o = &spatialmath.OrientationVectorDegrees{OZ: 1, Theta: heading}
+		theta := rdkutils.SwapCompassHeadingHandedness(headingLeft)
```
Comment (raybjork) [inline services/motion/localizer.go:149]: I don't think this is super necessary, the comment and the math make it pretty clear without it being a different function

File: services/motion/localizer.go [hunk 9]
```diff
@@ -56,8 +61,62 @@ type movementSensorLocalizer struct {
...
+	if math.IsNaN(gp.Lng()) {
+		return errors.New("lng can't be NaN")
+	}
+
+	return nil
+}
+
+// validatePose validates that a pose can be used for
+// motion planning.
+func validatePose(p spatialmath.Pose) error {
```
Comment (raybjork) [inline services/motion/localizer.go:92]: This should never be possible we control the construction of Poses pretty tightly, so I'm not a fan of adding this extra overhead in.

File: services/motion/localizer.go [hunk 10]
```diff
@@ -66,20 +125,29 @@ func (m *movementSensorLocalizer) CurrentPosition(ctx context.Context) (*referen
 	if err != nil {
 		return nil, err
 	}
+
+	if err := validateGeopoint(gp); err != nil {
```
Comment (raybjork) [inline services/motion/localizer.go:129]: Instead of this happening here I think this should just happen when we convert the protobuf to an RDK geopoint

**Review by raybjork — CHANGES_REQUESTED** [review]
We shouldn't need to do this validating in these locations and instead try to capture it as far upstream as possible.  NaN values should not be being passed into our functions if we include the validation as a part of construction of the RDK struct that is the counterpart to the protobuf

File: utils/math.go [hunk 11]
```diff
@@ -16,6 +16,12 @@ func DegToRad(degrees float64) float64 {
 	return degrees * math.Pi / 180
 }
 
+// SwapCompassHeadingHandedness converts left hand rule compass heading degrees
+// to right hand rule compass heading degrees & vice versa.
+func SwapCompassHeadingHandedness(heading float64) float64 {
```
[the remark this answers is not in the document]
Comment (biotinker) [reply utils/math.go:21]: +1 to this.

"Compass heading" is a left-handed value, period. If we convert to right-handedness, it's no longer compass heading, but some other thing that makes sense only in a particular context.

Having this function be more generally accessible might result in right-handed "compass headings" being passed around.
